include/sys/dmu_objset.h

```diff
@@ -125,6 +126,12 @@ struct objset {
 	kmutex_t os_user_ptr_lock;
 	void *os_user_ptr;
 	sa_os_t *os_sa;
+
+	/* kernel thread to upgrade this dataset */
+	kthread_t *os_upgrade_thread;
+	kmutex_t os_upgrade_lock;
+	kcondvar_t os_upgrade_cv;
+	boolean_t os_upgrade_exit;
```

Comment:
hmm.. I don't want to bind the upgrading thread or taskq to pool. This is supposed to be a system level tunables. I would rather have a module parameter to control the parallelism of upgrading, and then in ```dmu_objset_own()``` _and_ ```dmu_objset_sync()```we can spawn upgrade thread for this dataset if the current active upgrading threads is less than the parallelism parameter. Of course, we need to adjust the number of active thread when upgrade job is done for a dataset.

This seems much easier - do you see any problems from this?